keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,746 @@
+	// It means that we have container that pinned to the current NUMA node and all group nodes
+	Nodes []int `json:"nodes"`
+}
+
+// NodeMap contains memory information for each NUMA node.
+type NodeMap map[int]*NodeState
+
+// k8s.io/kubernetes/pkg/kubelet/cm/memorymanager/state
+package state
+
+// Block is a data structure used to represent a certain amount of memory
+type Block struct {
+	// NUMAAffinity contains the string that represents NUMA affinity bitmask
+	NUMAAffinity []int           `json:"numaAffinity"`
+	Type         v1.ResourceName `json:"type"`
+	Size         uint64          `json:"size"`
+}
+
+// ContainerMemoryAssignments stores memory assignments of containers
+type ContainerMemoryAssignments map[string]map[string][]Block
+
+// Reader interface used to read current memory/pod assignment state
+type Reader interface {
+	// GetMachineState returns Memory Map stored in the State
+	GetMachineState() NodeMap
+	// GetMemoryBlocks returns memory assignments of a container
+	GetMemoryBlocks(podUID string, containerName string) []Block
+	// GetMemoryAssignments returns ContainerMemoryAssignments
+	GetMemoryAssignments() ContainerMemoryAssignments
+}
+
+type writer interface {
+	// SetMachineState stores NodeMap in State
+	SetMachineState(memoryMap NodeMap)
+	// SetMemoryBlocks stores memory assignments of a container
+	SetMemoryBlocks(podUID string, containerName string, blocks []Block)
+	// SetMemoryAssignments sets ContainerMemoryAssignments by using the passed parameter
+	SetMemoryAssignments(assignments ContainerMemoryAssignments)
+	// Delete deletes corresponding Blocks from ContainerMemoryAssignments
+	Delete(podUID string, containerName string)
+	// ClearState clears machineState and ContainerMemoryAssignments
+	ClearState()
+}
+
+// State interface provides methods for tracking and setting memory/pod assignment
+type State interface {
+	Reader
+	writer
+}
+
+### Changes of existing components
+
+#### Container Manager changes
+
+Container Manager will create Memory Manager and register to Topology Manager as hint provider.
+
+#### Topology Manager changes
+
+Topology Manager will call out Memory Manager to gather topology hint and allocate memory resources during its admit sequence.
+
+#### Internal Container Lifecycle changes
+
+InternalContainerLifecycle will call out Memory Manager on container lifecycle events(Add/Stop) to allocate and reclaim memory resources.
+
+### Feature Gate and Kubelet Flags
+
+A new feature gate will be added to enable the Memory Manager feature. This feature gate will be disabled by default in the early releases.
+
+#### Feature Gate  
+`--feature-gate=MemoryManager=true`
+
+#### Memory Manager Policy Flag  
+`--memory-manager-policy` - to configure Memory Manager behaviour, `static` enables multi-NUMA (and single-NUMA) feature, whereas `none` disables it 
+
+Example:
+`--memory-manager-policy=none` or `--memory-manager-policy=static`
```

Comment:
I think we wanted to make it similar to the CPU manager, like static means that the container pinned to some NUMA nodes, but I do not have a strong opinion regarding the name, naming is always a problematic subject :)

Reply:
That's right, we followed the CPU Manager convention [1].
In terms of CPU affinity and pinning, words `statically` and `static` appears in the following context:

> As Guaranteed pods whose containers fit the requirements for being statically assigned are scheduled to the node, CPUs are removed from the shared pool and placed in the cpuset for the container. CFS quota is not used to bound the CPU usage of these containers as their usage is bound by the scheduling domain itself. In others words, the number of CPUs in the container cpuset is equal to the integer CPU limit specified in the pod spec. This static assignment increases CPU affinity and decreases context switches due to throttling for the CPU-bound workload.

In terms of memory, it can be seen in a similar way. Once memory is pinned to particular NUMA nodes, it does not change, i.e. it is `static`.

Besides, CPUs can be affined to multiple NUMA nodes by CPU Manager? So, the policy name `static` in CPU Manager could also be named `multi-numa`.

If the name of policy should be changed in MM from `static` to some other name, let us know.

[1] https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/